lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -13,45 +13,369 @@
+
+          auto out = wireOut(mulReg, op.instanceName() + "_out", b);
+          auto done = wireOut(doneReg, op.instanceName() + "_done", b);
+          wires.append({clk.input(), reset.input(), go.input(), left.input(),
+                        right.input(), out, done});
+        })
+        // Sequential operations.
+        .Case([&](RegisterOp op) {
+          auto in = wireIn(op.in(), op.instanceName() + "_in", b);
+          auto writeEn =
+              wireIn(op.write_en(), op.instanceName() + "_write_en", b);
+          auto clk = wireIn(op.clk(), op.instanceName() + "_clk", b);
+          auto reset = wireIn(op.reset(), op.instanceName() + "_reset", b);
+
+          auto outReg = reg(in, clk, reset, op.instanceName() + "_reg", b);
+          auto doneReg =
+              reg(writeEn, clk, reset, op.instanceName() + "_done_reg", b);
+
+          auto out = wireOut(outReg, op.instanceName(), b);
+          auto done = wireOut(doneReg, op.instanceName() + "_done", b);
+          wires.append({in.input(), writeEn.input(), clk.input(), reset.input(),
+                        out, done});
+        })
+        // Integer width modifying operations.
+        .Case([&](SliceLibOp op) {
+          auto in = wireIn(op.in(), op.instanceName() + "_in", b);
+          auto outWidth = op.out().getType().getIntOrFloatBitWidth();
+
+          auto extract = b.create<ExtractOp>(in, 0, outWidth);
+
+          auto out = wireOut(extract, op.instanceName() + "_out", b);
+          wires.append({in.input(), out});
+        })
+        // Structural operations.
+        .Case([&](WireLibOp op) {
+          auto wire = wireIn(op.in(), op.instanceName(), b);
+          wires.append({wire.input(), wire});
+        })
+        .Default([](Operation *) { return SmallVector<Value>(); });
+  }
+
+  template <typename OpTy, typename ResultTy>
+  void convertArithBinaryOp(OpTy op, SmallVectorImpl<Value> &wires,
+                            ImplicitLocOpBuilder &b) const {
+    auto left = wireIn(op.left(), op.instanceName() + "_left", b);
+    auto right = wireIn(op.right(), op.instanceName() + "_right", b);
+
+    auto add = b.create<ResultTy>(left, right);
+
+    auto out = wireOut(add, op.instanceName() + "_out", b);
+    wires.append({left.input(), right.input(), out});
+  }
+
+  template <typename OpTy>
+  void convertCompareBinaryOp(OpTy op, ICmpPredicate pred,
+                              SmallVectorImpl<Value> &wires,
+                              ImplicitLocOpBuilder &b) const {
+    auto left = wireIn(op.left(), op.instanceName() + "_left", b);
+    auto right = wireIn(op.right(), op.instanceName() + "_right", b);
+
+    auto add = b.create<ICmpOp>(pred, left, right);
+
+    auto out = wireOut(add, op.instanceName() + "_out", b);
+    wires.append({left.input(), right.input(), out});
+  }
+
+  ReadInOutOp wireIn(Value source, Twine name, ImplicitLocOpBuilder &b) const {
+    auto wire = b.create<WireOp>(source.getType(), name.str());
```

Comment:
Since we're always calling `wireIn/wireOut` with an underscored suffix, maybe that underscore should be included here?
```
auto wire = b.create<WireOp>(source.getType(), "_" + name.str());
```